File: apps/desktop/final_repo_updated/repo/apps/backend/app/api/routes/folders.py
"""API routes for folder operations.

These endpoints allow clients to create, retrieve, update and delete
folders, as well as list folders by parent. Folder metadata includes
icon information which may be used by the desktop application to
display custom folder icons.
"""

from __future__ import annotations

from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, status

from app.api.deps import get_current_active_db
from app.schemas.folder import FolderCreate, FolderResponse, FolderUpdate
from app.services import folder_service

router = APIRouter(prefix="/folders", tags=["folders"])


@router.get("/", response_model=List[FolderResponse])
def list_folders(
    *,
    db=Depends(get_current_active_db),
    parent_id: Optional[int] = None,
) -> List[FolderResponse]:
    """List folders filtered by their parent ID.

    Without a ``parent_id`` query parameter, only top‑level folders are
    returned. Providing a ``parent_id`` returns only folders whose
    parent matches the supplied ID. Recursive traversal is not
    supported; clients must request each level separately.
    """
    folders = folder_service.list_folders(db, parent_id=parent_id)
    return [FolderResponse.from_orm(f) for f in folders]


@router.get("/{folder_id}", response_model=FolderResponse)
def get_folder(*, db=Depends(get_current_active_db), folder_id: int) -> FolderResponse:
    """Retrieve a single folder by its ID."""
    folder = folder_service.get_folder(db, folder_id)
    return FolderResponse.from_orm(folder)


@router.post("/", response_model=FolderResponse, status_code=status.HTTP_201_CREATED)
def create_folder(*, db=Depends(get_current_active_db), folder: FolderCreate) -> FolderResponse:
    """Create a new folder.

    The parent folder must exist if ``parent_id`` is provided. No
    duplicate name checks are performed; clients should handle naming
    uniqueness according to their UI/UX guidelines.
    """
    if folder.parent_id is not None:
        # Ensure parent exists
        folder_service.get_folder(db, folder.parent_id)
    new_folder = folder_service.create_folder(db, folder)
    return FolderResponse.from_orm(new_folder)


@router.put("/{folder_id}", response_model=FolderResponse)
def update_folder(
    *,
    db=Depends(get_current_active_db),
    folder_id: int,
    folder: FolderUpdate,
) -> FolderResponse:
    """Update folder metadata and icon information."""
    updated = folder_service.update_folder(db, folder_id, folder)
    return FolderResponse.from_orm(updated)


@router.delete("/{folder_id}", status_code=status.HTTP_204_NO_CONTENT, response_model=None)
def delete_folder(*, db=Depends(get_current_active_db), folder_id: int) -> None:
    """Delete a folder and its children recursively."""
    folder_service.delete_folder(db, folder_id)

File: apps/desktop/src/api/folders.ts
import { request } from './client';
import type { Folder } from '../types/folder';

/**
 * List folders filtered by parent ID. If no parent ID is supplied the
 * backend returns only top‑level folders. Nested folders must be
 * requested separately.
 */
export async function listFolders(parentId?: number | null): Promise<Folder[]> {
  const query = parentId != null ? `?parent_id=${parentId}` : '';
  return await request<Folder[]>(`/folders/${query}`, {
    method: 'GET',
  });
}

/** Retrieve a single folder by ID. */
export async function getFolder(folderId: number): Promise<Folder> {
  return await request<Folder>(`/folders/${folderId}`, {
    method: 'GET',
  });
}

/** Create a new folder with the specified name and optional parent. */
export async function createFolder(
  name: string,
  parentId?: number | null,
): Promise<Folder> {
  const payload: any = { name };
  if (parentId != null) payload.parent_id = parentId;
  return await request<Folder>(`/folders/`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

/** Update an existing folder. Any of the provided fields will be
 * modified; omitted fields remain unchanged. To move a folder to the
 * root supply `null` for `parentId`.
 */
export async function updateFolder(
  folderId: number,
  { name, parentId }: { name?: string; parentId?: number | null },
): Promise<Folder> {
  const payload: any = {};
  if (name !== undefined) payload.name = name;
  if (parentId !== undefined) payload.parent_id = parentId;
  return await request<Folder>(`/folders/${folderId}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  });
}

/** Recursively delete a folder and all its children. */
export async function deleteFolder(folderId: number): Promise<void> {
  await request<void>(`/folders/${folderId}`, {
    method: 'DELETE',
  });
}
